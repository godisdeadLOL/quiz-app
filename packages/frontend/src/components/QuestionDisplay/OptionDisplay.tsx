import { Checkable } from "@/ui/Checkable";
import classNames from "classnames";
import type { QuestionPublic } from "shared/types";

type QuestionOptionDisplayProps = {
	question: QuestionPublic;
	index: number;
	isPicked: boolean;
	onChangePicked?: (value: boolean) => void;
};

export function OptionDisplay({ question, index, isPicked, onChangePicked: onPick }: QuestionOptionDisplayProps) {
	return (
		<label className={classNames("button flex gap-4 px-4 py-2 rounded-lg", isPicked ? "bg-surface-2" : "bg-surface-1")}>
			<div className="flex-1">{question.options[index].description}</div>
			<Checkable value={isPicked} type={question.mode === "single" ? "radio" : "checkbox"} onChangeValue={onPick} />
		</label>
	);
}
