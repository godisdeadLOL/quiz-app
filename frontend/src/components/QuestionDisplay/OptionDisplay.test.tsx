import type { QuestionPublic } from "@/api/types";
import { cleanup, render, screen } from "@testing-library/react";

import { OptionDisplay } from "./OptionDisplay";

const question: QuestionPublic = {
	description: "описание",
	mode: "single",
	options: [{ description: "Вариант 1" }, { description: "Вариант 2" }, { description: "Вариант 3" }],
};

describe("<OptionDisplay>", () => {
	function testOption(index: number, mode: QuestionPublic["mode"], isPicked: boolean) {
		render(<OptionDisplay question={{ ...question, mode }} index={index} isPicked={isPicked} />);

		const root = screen.getByTestId(`option-${index}`);
		const checkable = screen.getByTestId("checkable");
		const checkableInput = screen.getByTestId("checkable").querySelector("input")!;

		expect(root).toHaveTextContent(question.options[index].description);
		expect(checkableInput).toHaveAttribute("type", mode === "single" ? "radio" : "checkbox");
		expect(checkable.getAttribute("aria-checked")).toBe(String(isPicked));
	}

	test("mode=single, isPicked=false", () => testOption(0, "single", false));
	test("mode=multiple, isPicked=true", () => testOption(1, "multiple", true));
	test("mode=single, isPicked=true", () => testOption(2, "single", true));

	afterEach(() => cleanup());
});
