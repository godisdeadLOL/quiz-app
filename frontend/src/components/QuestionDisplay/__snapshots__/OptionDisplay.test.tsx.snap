// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test 1`] = `
<label
  class="button flex gap-4 px-4 py-2 rounded-lg bg-surface-1"
  data-testid="option-0"
>
  <div
    class="flex-1"
  >
    Вариант 1
  </div>
  <label>
    <input
      hidden=""
      type="radio"
    />
    <div
      class="box-content size-[14px] p-1 text-sm border-1 text-white border-border rounded-full"
      data-checked="false"
    />
  </label>
</label>
`;
